device_test.go

```diff
@@ -227,20 +217,54 @@ func (*failDetachLib) detachKernelDriver(h *libusbDevHandle, i uint8) error {
 
 func TestAutoDetachFailure(t *testing.T) {
 	t.Parallel()
-	fake := newFakeLibusb()
-	c := newContextWithImpl(&failDetachLib{fake})
-	defer c.Close()
-	dev, err := c.OpenDeviceWithVIDPID(0x8888, 0x0002)
-	if dev == nil {
-		t.Fatal("OpenDeviceWithVIDPID(0x8888, 0x0002): got nil device, need non-nil")
-	}
-	defer dev.Close()
+	dev, err := createFakeDevice(0x8888, 0x0002)
 	if err != nil {
-		t.Fatalf("OpenDeviceWithVIDPID(0x8888, 0x0002): %v", err)
+		t.Error(err)
 	}
 	dev.SetAutoDetach(true)
 	_, err = dev.Config(1)
 	if err == nil {
 		t.Fatalf("%s.Config(1) got nil, but want no nil because interface fails to detach", dev)
 	}
 }
+
+func TestActiveConfigNumFailure(t *testing.T) {
+	dev, err := createFakeDevice(0x8888, 0x0002)
+	if err != nil {
+		t.Error(err)
+	}
+
+	dev.handle = nil
+	_, err = dev.ActiveConfigNum()
+	if err == nil {
+		t.Fatalf("Error on failing ActiveConfigNum()")
+	}
+}
+
+func TestDeviceDescString(t *testing.T) {
+	dev, err := createFakeDevice(0x8888, 0x0002)
+	if err != nil {
+		t.Error(err)
+	}
+
+	expected := "1.2: 8888:0002 (available configs: [1])"
+	result := dev.Desc.String()
+
+	if expected != result {
+		t.Fatalf("Device String() failed, expeceted: \"%s\", got: \"%s\"", expected, result)
+	}
+}
+
+func createFakeDevice(vid ID, pid ID) (*Device, error) {
+	fake := newFakeLibusb()
+	c := newContextWithImpl(&failDetachLib{fake})
+	dev, err := c.OpenDeviceWithVIDPID(vid, pid)
+	if dev == nil {
+		return nil, err
+	}
```

Comment:
you should not be testing for `dev == nil` as the first check, because if err is != nil, dev is undefined. It may very well be that dev is not nil, but because err is != nil, dev is unusable.
Check for err != nil instead.